### PR TITLE
Feat/new stock

### PR DESCRIPTION
**Pull request by LinkageAssPlc**
captured new stock

### CODE REVIEW

File: src/inventory-brand/api/controllers/brand.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import { BrandDTO } from "../../DTOs/BrandDTO";
+import { BrandService } from "../../services";
+import { BaseController } from "../../../inventory-shared/api";
+
+
+export class BrandController{
+    static brand = BaseController(async (request: Request) => {
+        const BrandDTO = request.body as BrandDTO;
+        BrandDTO.userID = new Types.ObjectId(request.token._id);
+        const{status, message, data} = await BrandService(BrandDTO)
```
Comment (TimiBolu) [inline src/inventory-brand/api/controllers/brand.ts:13]: this should be addBrandService not brandService

File: src/inventory-brand/api/controllers/brand.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import { Types } from "mongoose";
+import { Request } from "express";
+
+import { BrandDTO } from "../../DTOs/BrandDTO";
+import { BrandService } from "../../services";
+import { BaseController } from "../../../inventory-shared/api";
+
+
+export class BrandController{
+    static brand = BaseController(async (request: Request) => {
```
Comment (TimiBolu) [inline src/inventory-brand/api/controllers/brand.ts:10]: this should be addBrand not brand

File: src/inventory-brand/api/routes/index.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import CategoryValidation from "../validations";
+import { baseRouter, baseValidation } from "../../../inventory-shared/api";
+
+import { BrandController } from "../controllers/brand";
+import { AuthMiddleware } from "../../../inventory-auth/middlewares/authMiddleware";
+
+const {POST, router} = baseRouter();
+
+POST("/add-brand", [baseValidation(CategoryValidation.brand), AuthMiddleware.baseAuthToken, AuthMiddleware.IsUserMiddleware, BrandController.brand]);
```
Comment (TimiBolu) [inline src/inventory-brand/api/routes/index.ts:9]: this should be "/" not "/add-brand" as create operations for POST are commonly named

File: src/inventory-brand/DTOs/BrandDTO.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { ObjectId } from "mongodb";
+
+
+export type BrandDTO = {
+    userID: ObjectId;
+    name: String;
```
Comment (TimiBolu) [inline src/inventory-brand/DTOs/BrandDTO.ts:6]: this shoulde be string not String. "String" is a string constructor. "string" is a type.

File: src/inventory-brand/api/controllers/brand.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { Types } from "mongoose";
+import { Request } from "express";
+
+import { BrandDTO } from "../../DTOs/BrandDTO";
```
Comment (TimiBolu) [inline src/inventory-brand/api/controllers/brand.ts:4]: This should be AddBrandDTO not BrandDTO

File: src/inventory-category/api/controllers/category.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+import { CategoryDTO } from "../../DTOs/CategoryDTO";
+import { CategoryService } from "../../services";
+import { BaseController } from "../../../inventory-shared/api";
+
+
+export class CategoryController{
+    static category = BaseController(async (request: Request) => {
+        const CategoryDTO = request.body as CategoryDTO;
+        console.log(CategoryDTO)
```
Comment (TimiBolu) [inline src/inventory-category/api/controllers/category.ts:12]: remove all logs except they are for error logging

File: src/inventory-category/api/controllers/category.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+        CategoryDTO.userID = new Types.ObjectId(request.token._id);
+        console.log(CategoryDTO.userID);
+        const{status, message, data} = await CategoryService(CategoryDTO)
```
Comment (TimiBolu) [inline src/inventory-category/api/controllers/category.ts:15]: this should be AddCategoryService

File: src/inventory-category/api/controllers/category.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import { Types } from "mongoose";
+import { Request } from "express";
+
+import { CategoryDTO } from "../../DTOs/CategoryDTO";
+import { CategoryService } from "../../services";
+import { BaseController } from "../../../inventory-shared/api";
+
+
+export class CategoryController{
+    static category = BaseController(async (request: Request) => {
```
Comment (TimiBolu) [inline src/inventory-category/api/controllers/category.ts:10]: this should be AddCategoryDTO

File: src/inventory-category/api/routes/index.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import CategoryValidation from "../validations";
+import { baseRouter, baseValidation } from "../../../inventory-shared/api";
+
+import { CategoryController } from "../controllers/category";
+import { AuthMiddleware } from "../../../inventory-auth/middlewares/authMiddleware";
+
+const {POST, router} = baseRouter();
+
+POST("/add-category", [baseValidation(CategoryValidation.category), AuthMiddleware.baseAuthToken, AuthMiddleware.IsUserMiddleware, CategoryController.category]);
```
Comment (TimiBolu) [inline src/inventory-category/api/routes/index.ts:9]: this should be "/" not add-category

File: src/inventory-category/api/controllers/category.ts (hunk 8 above)
Comment (TimiBolu) [inline src/inventory-category/api/controllers/category.ts:10]: this should be static addCategory not static category

File: src/inventory-entities/NewStock.ts [hunk 10]
```diff
@@ -1,37 +1,34 @@
...
-            category: {type: String, required: true},
-            brand: {type: String, required: true},
+            productID: {type: ObjectId, required: true, ref: ModelNames.PRODUCT},
+            categoryID: {type: ObjectId, required: true, ref: ModelNames.CATEGORY},
+            brandID: {type: ObjectId, required: true, ref: ModelNames.BRAND},
             quantity: {type: String, required: true},
             price: {type: Number, required: true},
-            unit: {type: String, required: true},
-            isInStock: {type: Boolean, default: false}
+            unit: {type: String, enum: [Units.UNIT, Units.CARTONS, Units.GALLONS, Units.PACK, Units.PIECES, Units.RIMS], default: Units.UNIT, required: true},
```
Comment (TimiBolu) [inline src/inventory-entities/NewStock.ts:30]: use Object.values(Units) instead. It's more helpful as you change the Units enum. You won't have to manually update this

File: src/inventory-new-stock/api/controllers/new-stock.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import { Types } from "mongoose";
+import { Request } from "express";
+
+import { NewStockDTO } from "../../DTOs/NewStockDTO";
+import { NewStockService } from "../../services";
+import { BaseController } from "../../../inventory-shared/api";
+
+
+export class NewStockController{
+    static newStock = BaseController(async (request: Request) => {
```
Comment (TimiBolu) [inline src/inventory-new-stock/api/controllers/new-stock.ts]: should be addNewStock

File: src/inventory-new-stock/api/routes/index.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import ProductValidation from "../validations";
+import { baseRouter, baseValidation } from "../../../inventory-shared/api";
+
+import { NewStockController } from "../controllers/new-stock";
+import { AuthMiddleware } from "../../../inventory-auth/middlewares/authMiddleware";
+
+const {POST, router} = baseRouter();
+
+POST("/add-new-stock", [baseValidation(ProductValidation.newStock), AuthMiddleware.baseAuthToken, AuthMiddleware.IsUserMiddleware, NewStockController.newStock]);
```
Comment (TimiBolu) [inline src/inventory-new-stock/api/routes/index.ts]: shoule bde "/" not "/add-new-stock"

File: src/inventory-new-stock/services/new-stock.service.ts [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    // let product: Product;
+
+    await Promise.all(
+        itemList.map(
+            async(item, index) => {
+                const productExist = await ProductModel.findById(item.productID) as Product
+                const categoryExist = await CategoryModel.findById(item.categoryID) as Category
+                const brandExist = await BrandModel.findById(item.brandID) as Brand
+            
+                if (!productExist || !categoryExist || !brandExist) {
```
Comment (TimiBolu) [inline src/inventory-new-stock/services/new-stock.service.ts]: this condition is redundant

File: src/inventory-new-stock/services/new-stock.service.ts [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+import { NewStockDTO } from "../DTOs/NewStockDTO";
+import { getUser } from "../../inventory-accounts/user/services";
+import { NewStockModel } from "../../inventory-entities/NewStock";
+import { ProductModel } from "../../inventory-entities/Product";
+import { CategoryModel } from "../../inventory-entities/Category";
+import { BrandModel } from "../../inventory-entities/Brand";
+import { Brand, Category, Product } from "../../types/user";
+
+
+export const NewStockService = async ({userID, itemList}: NewStockDTO) => {
```
Comment (TimiBolu) [inline src/inventory-new-stock/services/new-stock.service.ts]: should be AddNewCategoryService

File: src/inventory-new-stock/api/validations/index.ts [hunk 15]
```diff
@@ -0,0 +1,24 @@
+import { Joi } from "celebrate";
+
+import { Units } from "../../../types/user";
+import { toObjectId } from "../../../inventory-shared/validateAndConvertToObjectID";
+
+
+export default{
+    newStock: {
```
Comment (TimiBolu) [inline src/inventory-new-stock/api/validations/index.ts:8]: addNewStock

File: src/inventory-product/api/controllers/product.ts [hunk 16]
```diff
@@ -0,0 +1,16 @@
+import { Types } from "mongoose";
+import { Request } from "express";
+
+import { ProductDTO } from "../../DTOs/ProductDTO";
+import { ProductService } from "../../services";
+import { BaseController } from "../../../inventory-shared/api";
+
+
+export class ProductController{
+    static product = BaseController(async (request: Request) => {
```
Comment (TimiBolu) [inline src/inventory-product/api/controllers/product.ts:10]: addProduct

File: src/inventory-product/api/routes/index.ts [hunk 17]
```diff
@@ -0,0 +1,11 @@
+import ProductValidation from "../validations";
+import { baseRouter, baseValidation } from "../../../inventory-shared/api";
+
+import { ProductController } from "../controllers/product";
+import { AuthMiddleware } from "../../../inventory-auth/middlewares/authMiddleware";
+
+const {POST, router} = baseRouter();
+
+POST("/add-product", [baseValidation(ProductValidation.product), AuthMiddleware.baseAuthToken, AuthMiddleware.IsUserMiddleware, ProductController.product]);
```
Comment (TimiBolu) [inline src/inventory-product/api/routes/index.ts:9]: should be "/" not "/add-product"

**Comment by LinkageAssPlc** [comment]
all recommendations have been done

**Pull request merged**